### ISSUE TITLE
Issue with spinning up theforeman.example.com

### ISSUE THREAD

**Issue by jhescobar05** [garystafford/foreman-vagrant#5]
Failed to provision  Foreman VM first. 

Virtualbox  5.0.2
Vagrant 1.7.4

Error Message: 

==> theforeman.example.com: /usr/lib/ruby/site_ruby/1.8/rubygems/custom_require.rb:31:in `gem_original_require'
==> theforeman.example.com: : no such file to load -- puppet (LoadError)
==> theforeman.example.com:     from /usr/lib/ruby/site_ruby/1.8/rubygems/custom_require.rb:31:in`require'
==> theforeman.example.com:     from /usr/lib/ruby/gems/1.8/gems/kafo_parsers-0.0.5/lib/kafo_parsers/puppet_module_parser.rb:2
==> theforeman.example.com:     from /usr/lib/ruby/site_ruby/1.8/rubygems/custom_require.rb:31:in `gem_original_require'
==> theforeman.example.com:     from /usr/lib/ruby/site_ruby/1.8/rubygems/custom_require.rb:31:in`require'
==> theforeman.example.com:     from /usr/lib/ruby/gems/1.8/gems/kafo-0.6.6/lib/kafo/puppet_module.rb:4
==> theforeman.example.com:     from /usr/lib/ruby/site_ruby/1.8/rubygems/custom_require.rb:31:in `gem_original_require'
==> theforeman.example.com:     from /usr/lib/ruby/site_ruby/1.8/rubygems/custom_require.rb:31:in`require'
==> theforeman.example.com:     from /usr/lib/ruby/gems/1.8/gems/kafo-0.6.6/lib/kafo/configuration.rb:4
==> theforeman.example.com:     from /usr/lib/ruby/site_ruby/1.8/rubygems/custom_require.rb:31:in `gem_original_require'
==> theforeman.example.com:     from /usr/lib/ruby/site_ruby/1.8/rubygems/custom_require.rb:31:in`require'
==> theforeman.example.com:     from /usr/lib/ruby/gems/1.8/gems/kafo-0.6.6/lib/kafo/kafo_configure.rb:14
==> theforeman.example.com:     from /usr/lib/ruby/site_ruby/1.8/rubygems/custom_require.rb:31:in `gem_original_require'
==> theforeman.example.com:     from /usr/lib/ruby/site_ruby/1.8/rubygems/custom_require.rb:31:in`require'
==> theforeman.example.com:     from /usr/lib/ruby/gems/1.8/gems/kafo-0.6.6/lib/kafo.rb:3
==> theforeman.example.com:     from /usr/lib/ruby/site_ruby/1.8/rubygems/custom_require.rb:36:in `gem_original_require'
==> theforeman.example.com:     from /usr/lib/ruby/site_ruby/1.8/rubygems/custom_require.rb:36:in`require'
==> theforeman.example.com:     from /usr/sbin/foreman-installer:3

**Comment by snemetz** [comment]
I just ran into the same thing. Also, sudo asked for a password (which failed) early on near the 'yum update'. Trying to figure out

**Comment by snemetz** [comment]
gem install puppet --version '< 4' --no-rdoc --no-ri
Appears to fix this. 'yum install puppet-server-3.8.2-1' should also work.
Now got an error about augeas. Guessing it didn't get installed either.

**Comment by snemetz** [comment]
Ok, problem appears to be Puppet 4
To get foreman-installer working after vagrant brings up the VM:
yum erase -y puppet-agent
rm -f /etc/yum.repos.d/puppetlabs-pc1.repo
yum clean all
yum install -y puppet-server
yum install -y foreman-installer
foreman-installer

Need to look at vagrant setup to make the changes in provisioning script

**Comment by jhescobar05** [comment]
You're right about the puppet version being installed on the box. I updated my vagrant file to pull 'puppetlabs/centos-6.6-64-puppet-enterprise' since that box has puppet 3.8.1 installed.  The one that's currently being pulled has Puppet 4.2.1 -> [puppetlabs/centos-6.6-64-puppet](https://atlas.hashicorp.com/puppetlabs/boxes/centos-6.6-64-puppet). 

You have to use  the password  you use to log into your host. Wasn't happy about that part. 

I'm running this set up on a mac so I'm having issues trying to view the foreman page.

**Comment by snemetz** [comment]
I'm also doing this on a mac. I added port forwarding to the node file and Vagrant say it configured the forwarding, but it is not working for some reason. Will be trying to figure it out tomorrow.

**Comment by jhescobar05** [comment]
Haven't had a chance to look into port forwarding issues. [http://salferrarello.com/mac-pfctl-port-forwarding/](http://salferrarello.com/mac-pfctl-port-forwarding/)

**Comment by garystafford** [comment]
I also now get the same errors when re-building the project as is, from scratch, as well as the sudo password request. I also tried using the later centos-7 box and later version of foreman-1.9, which resulted in same errors:

```
==> theforeman.example.com: --> Processing Dependency: puppet < 4.0.0 for package: rubygem-kafo-0.6.11-1.el7.noarch
==> theforeman.example.com: Package puppet-3.8.2-1.el7.noarch is obsoleted by puppet-agent-1.2.2-1.el7.x86_64 which is already installed
==> theforeman.example.com: --> Finished Dependency Resolution
==> theforeman.example.com:  You could try using --skip-broken to work around the problem
==> theforeman.example.com: Error: Package: rubygem-kafo-0.6.11-1.el7.noarch (foreman)
==> theforeman.example.com:            Requires: puppet < 4.0.0
==> theforeman.example.com:            Installed: puppet-agent-1.2.2-1.el7.x86_64 (@puppetlabs-pc1)
==> theforeman.example.com:                puppet >= 4.0.0-1.el7
==> theforeman.example.com:            Available: puppet-3.5.1-1.el7.noarch (puppetlabs-products)
==> theforeman.example.com:                puppet = 3.5.1-1.el7
==> theforeman.example.com:            Available: puppet-3.6.0-1.el7.noarch (puppetlabs-products)
...
==> theforeman.example.com:  You could try running: rpm -Va --nofiles --nodigest
==> theforeman.example.com: sudo: puppet: command not found
...
==> theforeman.example.com: sudo: puppet: command not found
==> theforeman.example.com: sudo: puppet: command not found
==> theforeman.example.com: sudo: puppet: command not found
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.
```

**Comment by jhescobar05** [comment]
Hey Gary,
The box that is being pulled from atlas has puppet 4.2.1 installed.  

I wasn't able to go through the set up due to my Mac issue. I don't know what else will be needed. 

Thank you for providing a pretty cool example.

**Comment by garystafford** [comment]
I will work on an update as version dependencies align better. We use foreman and I would like to maintain a working installer. Thanks for all the input on this issue.

**Comment by snemetz** [comment]
I forked the repo and changed the provisioning script to remove the puppet agent and puppet 4.0 repo, before installing foreman installer.

The port forwarding issue is a firewalling issue. So, I disabled the firewall for now. As I'm more concerned with doing some testing then the security aspect at the moment.

Have it up and running now :)

**Comment by jhescobar05** [comment]
Thanks For your input.

**Comment by garystafford** [comment]
@snemetz (@jhescobar05) I know you are working on a fix with some other improvements in a fork of main branch, so I created a second branch (https://github.com/garystafford/foreman-vagrant/tree/centos_7), and:
1. Fixed the main issue, based on your suggestion to re-deploy pre-4.x version of Puppet to all nodes (master and agents)
2. Fixed the sudo/password issue
3. Fixed firewall issues in CentOS 7 with use of `firewalld` vs. `iptables`
4. Switched to CentOS 7 from CentOS 6.x

I ran it few times, and seems to all work. Registered certificates from nodes, etc. I have to run some more tests, but appeared to all be working.
